File: mobile/src/services/realTimePreviewService.ts
// ✅ NOUVEAU: Service de preview temps réel pendant l'édition vidéo

import { TimelineScene, VideoTimeline } from '../components/TimelinePreview';

export interface EffectConfig {
    name: string;
    intensity?: number; // 0.0 à 1.0
    parameters?: Record<string, any>;
}

export interface TransitionConfig {
    type: 'fade' | 'slide' | 'zoom' | 'wipe' | 'dissolve' | 'none';
    duration?: number; // secondes
}

export interface RealtimePreviewParams {
    timeline: VideoTimeline;
    currentTime: number;
    effects?: EffectConfig[];
    transitions?: TransitionConfig[];
    playbackRate?: number;
}

export interface EffectShader {
    name: string;
    apply: (frame: ImageData, intensity: number, params?: Record<string, any>) => ImageData;
    webglShader?: string; // Pour extension WebGL future
}

/**
 * Service de preview temps réel qui applique les effets/transitions en temps réel
 * sur le device sans attendre un rendu backend
 */
export class RealTimePreviewService {
    private effectCache: Map<string, EffectShader> = new Map();
    private previewBuffer: Map<string, HTMLVideoElement | null> = new Map();

    constructor() {
        this.initializeEffects();
    }

    /**
     * Initialise les shaders d'effets disponibles
     */
    private initializeEffects(): void {
        // Effets de base - peuvent être étendus avec WebGL shaders
        const effects: EffectShader[] = [
            {
                name: 'zoom',
                apply: (frame, intensity) => frame, // Placeholder - sera implémenté avec transformation
            },
            {
                name: 'fade',
                apply: (frame, intensity) => frame,
            },
            {
                name: 'glow',
                apply: (frame, intensity) => frame,
            },
            {
                name: 'blur',
                apply: (frame, intensity) => frame,
            },
            {
                name: 'sharpen',
                apply: (frame, intensity) => frame,
            },
            {
                name: 'vintage',
                apply: (frame, intensity) => frame,
            },
            {
                name: 'neon',
                apply: (frame, intensity) => frame,
            },
            {
                name: 'blackwhite',
                apply: (frame, intensity) => frame,
            },
            {
                name: 'warm',
                apply: (frame, intensity) => frame,
            },
            {
                name: 'cool',
                apply: (frame, intensity) => frame,
            },
        ];

        effects.forEach(effect => {
            this.effectCache.set(effect.name, effect);
        });
    }

    /**
     * Obtient les paramètres d'effets pour une position dans la timeline
     * Retourne les paramètres sans la vidéo (pour calcul local)
     */
    getEffectParams(params: RealtimePreviewParams): {
        sceneIndex: number;
        scene: TimelineScene;
        activeEffects: EffectConfig[];
        transition: TransitionConfig | null;
        shouldApplyTransition: boolean;
    } {
        const { timeline, currentTime } = params;

        // Trouver la scène active
        let activeSceneIndex = 0;
        let activeScene = timeline.scenes[0];

        for (let i = 0; i < timeline.scenes.length; i++) {
            const scene = timeline.scenes[i];
            if (currentTime >= scene.start_time && currentTime < scene.start_time + scene.duration) {
                activeSceneIndex = i;
                activeScene = scene;
                break;
            }
        }

        // Effets actifs de la scène
        const activeEffects: EffectConfig[] = (activeScene.effects || [])
            .map(effectName => ({
                name: effectName,
                intensity: 1.0,
            }));

        // Transition entre scènes
        let transition: TransitionConfig | null = null;
        let shouldApplyTransition = false;

        if (activeScene.transition && activeScene.transition !== 'none') {
            const transitionDuration = 0.5; // Durée par défaut
            const sceneEndTime = activeScene.start_time + activeScene.duration;
            const transitionStartTime = sceneEndTime - transitionDuration;

            if (currentTime >= transitionStartTime) {
                transition = {
                    type: activeScene.transition as any,
                    duration: transitionDuration,
                };
                shouldApplyTransition = true;
            }
        }

        return {
            sceneIndex: activeSceneIndex,
            scene: activeScene,
            activeEffects,
            transition,
            shouldApplyTransition,
        };
    }

    /**
     * Prépare le preview pour scrubbing fluide
     * Cache les médias nécessaires pour performance
     */
    async preparePreviewBuffer(timeline: VideoTimeline): Promise<void> {
        // Précharger les médias de la timeline pour performance
        const mediaUrls = timeline.scenes
            .filter(scene => scene.media_url)
            .map(scene => scene.media_url!)
            .filter((url, index, self) => self.indexOf(url) === index); // Unique

        for (const url of mediaUrls) {
            if (!this.previewBuffer.has(url)) {
                try {
                    // Précharger la vidéo (sera utilisé par expo-video)
                    // Pour l'instant, on marque juste comme chargé
                    this.previewBuffer.set(url, null);
                } catch (error) {
                    console.warn(`[RealTimePreview] Impossible de précharger ${url}:`, error);
                }
            }
        }
    }

    /**
     * Nettoie le buffer de preview
     */
    clearPreviewBuffer(): void {
        this.previewBuffer.clear();
    }

    /**
     * Vérifie si un effet est disponible
     */
    isEffectAvailable(effectName: string): boolean {
        return this.effectCache.has(effectName);
    }

    /**
     * Liste tous les effets disponibles
     */
    getAvailableEffects(): string[] {
        return Array.from(this.effectCache.keys());
    }
}

// Instance singleton
let previewServiceInstance: RealTimePreviewService | null = null;

/**
 * Obtient l'instance singleton du service de preview temps réel
 */
export const getRealTimePreviewService = (): RealTimePreviewService => {
    if (!previewServiceInstance) {
        previewServiceInstance = new RealTimePreviewService();
    }
    return previewServiceInstance;
};

/**
 * Service helper pour intégration facile
 */
export const realTimePreviewService = {
    /**
     * Obtient les paramètres d'effets pour une position dans la timeline
     */
    getEffectParams: (params: RealtimePreviewParams) => {
        const service = getRealTimePreviewService();
        return service.getEffectParams(params);
    },

    /**
     * Prépare le preview pour scrubbing fluide
     */
    preparePreview: async (timeline: VideoTimeline) => {
        const service = getRealTimePreviewService();
        await service.preparePreviewBuffer(timeline);
    },

    /**
     * Nettoie le buffer
     */
    clearBuffer: () => {
        const service = getRealTimePreviewService();
        service.clearPreviewBuffer();
    },

    /**
     * Vérifie si un effet est disponible
     */
    isEffectAvailable: (effectName: string): boolean => {
        const service = getRealTimePreviewService();
        return service.isEffectAvailable(effectName);
    },

    /**
     * Liste tous les effets disponibles
     */
    getAvailableEffects: (): string[] => {
        const service = getRealTimePreviewService();
        return service.getAvailableEffects();
    },
};

File: backend/src/services/realtime_preview_service.rs
// ✅ NOUVEAU: Service de preview temps réel pour calcul local côté client

use crate::core::types::{AppError, AppResult};
use log::info;
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RealtimePreviewRequest {
    pub timeline: Value, // VideoTimeline JSON
    pub current_time: f64,
    pub effects: Option<Vec<String>>,
    pub transitions: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EffectParam {
    pub name: String,
    pub intensity: Option<f32>,
    pub parameters: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransitionParam {
    pub r#type: String,
    pub duration: Option<f32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RealtimePreviewResponse {
    pub success: bool,
    pub scene_index: usize,
    pub scene: Value,
    pub active_effects: Vec<EffectParam>,
    pub transition: Option<TransitionParam>,
    pub should_apply_transition: bool,
    pub media_url: Option<String>,
    pub timestamp: f64,
}

/// Service de preview temps réel qui retourne les paramètres d'effets
/// pour calcul local (pas la vidéo elle-même)
pub struct RealtimePreviewService;

impl RealtimePreviewService {
    pub fn new() -> Self {
        Self
    }

    /// Obtient les paramètres d'effets pour une position dans la timeline
    /// Retourne les paramètres sans la vidéo (pour calcul local)
    pub fn get_effect_params(
        &self,
        request: RealtimePreviewRequest,
    ) -> AppResult<RealtimePreviewResponse> {
        info!(
            "[RealtimePreview] Calcul paramètres effets - current_time: {}s",
            request.current_time
        );

        // Parser la timeline
        let timeline = request.timeline;
        let scenes = timeline.get("scenes").and_then(|s| s.as_array()).ok_or_else(|| {
            AppError::BadRequest("Timeline invalide: 'scenes' manquant".to_string())
        })?;

        // Trouver la scène active
        let mut active_scene_index = 0;
        let mut active_scene = None;

        for (index, scene) in scenes.iter().enumerate() {
            let start_time = scene.get("start_time").and_then(|s| s.as_f64()).unwrap_or(0.0);
            let duration = scene.get("duration").and_then(|d| d.as_f64()).unwrap_or(0.0);

            if request.current_time >= start_time && request.current_time < start_time + duration {
                active_scene_index = index;
                active_scene = Some(scene.clone());
                break;
            }
        }

        let active_scene = active_scene.ok_or_else(|| {
            AppError::BadRequest("Aucune scène trouvée pour le temps spécifié".to_string())
        })?;

        // Extraire les effets de la scène
        let scene_effects = active_scene
            .get("effects")
            .and_then(|e| e.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|v| v.as_str().map(|s| s.to_string()))
                    .collect::<Vec<String>>()
            })
            .unwrap_or_default();

        let active_effects: Vec<EffectParam> = scene_effects
            .iter()
            .map(|effect_name| EffectParam {
                name: effect_name.clone(),
                intensity: Some(1.0),
                parameters: None,
            })
            .collect();

        // Extraire la transition
        let transition_type = active_scene
            .get("transition")
            .and_then(|t| t.as_str())
            .unwrap_or("none")
            .to_string();

        let transition_duration = 0.5; // Durée par défaut
        let scene_start_time =
            active_scene.get("start_time").and_then(|s| s.as_f64()).unwrap_or(0.0);
        let scene_duration = active_scene.get("duration").and_then(|d| d.as_f64()).unwrap_or(0.0);
        let scene_end_time = scene_start_time + scene_duration;
        let transition_start_time = scene_end_time - transition_duration as f64;

        let (transition, should_apply_transition) =
            if transition_type != "none" && request.current_time >= transition_start_time {
                (
                    Some(TransitionParam {
                        r#type: transition_type,
                        duration: Some(transition_duration),
                    }),
                    true,
                )
            } else {
                (None, false)
            };

        // Extraire l'URL média
        let media_url =
            active_scene.get("media_url").and_then(|u| u.as_str()).map(|s| s.to_string());

        Ok(RealtimePreviewResponse {
            success: true,
            scene_index: active_scene_index,
            scene: active_scene.clone(),
            active_effects,
            transition,
            should_apply_transition,
            media_url,
            timestamp: request.current_time,
        })
    }
}

impl Default for RealtimePreviewService {
    fn default() -> Self {
        Self::new()
    }
}
